sync/settings.go

```diff
@@ -0,0 +1,305 @@
+		}
+		settings = config.Flatten(p.Config)
+	}
+
+	// Check if there any changed settings.
+	if len(settings) == 0 {
+		return nil, ErrUnchanged
+	}
+
+	// Expand config to hierarchical form.
+	settings = config.Expand(settings)
+
+	return &SettingsExport{
+		Type:   TypeSettings,
+		Config: settings,
+	}, nil
+}
+
+// ImportSettings imports the global settings.
+func ImportSettings(r *SettingsImportRequest) (*ImportResult, error) {
+	// Check import.
+	if r.Export.Type != TypeSettings {
+		return nil, ErrMismatch
+	}
+
+	// Flatten config.
+	settings := config.Flatten(r.Export.Config)
+
+	// Validate config and gather some metadata.
+	var (
+		result  = &ImportResult{}
+		checked int
+	)
+	err := config.ForEachOption(func(option *config.Option) error {
+		// Check if any setting is set.
+		if r.Reset && option.IsSetByUser() {
+			result.ReplacesExisting = true
+		}
+
+		newValue, ok := settings[option.Key]
+		if ok {
+			checked++
+
+			// Validate the new value.
+			if err := option.ValidateValue(newValue); err != nil {
+				return fmt.Errorf("%w: configuration value for %s is invalid: %s", ErrInvalidSetting, option.Key, err)
+			}
+
+			// Collect metadata.
+			if option.RequiresRestart {
+				result.RestartRequired = true
+			}
+			if !r.Reset && option.IsSetByUser() {
+				result.ReplacesExisting = true
+			}
+		}
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+	if checked < len(settings) {
+		return nil, fmt.Errorf("%w: the export contains unknown settings", ErrInvalidImport)
+	}
+
+	// Import global settings.
+	if r.Target == ExportTargetGlobal {
+		// Stop here if we are only validating.
+		if r.ValidateOnly {
+			return result, nil
+		}
+
+		// Import to global config.
+		vErrs, restartRequired := config.ReplaceConfig(settings)
+		if len(vErrs) > 0 {
+			s := make([]string, 0, len(vErrs))
+			for _, err := range vErrs {
+				s = append(s, err.Error())
+			}
+			return nil, fmt.Errorf(
+				"%w: the supplied configuration could not be applied:\n%s",
+				ErrImportFailed,
+				strings.Join(s, "\n"),
+			)
+		}
+
+		result.RestartRequired = restartRequired
+		return result, nil
+	}
+
+	// Import settings into profile.
+	rec, err := db.Get(profile.ProfilesDBPath + r.Target)
+	if err != nil {
+		return nil, fmt.Errorf("%w: failed to find profile: %s", ErrTargetNotFound, err)
+	}
+	p, err := profile.EnsureProfile(rec)
+	if err != nil {
+		return nil, fmt.Errorf("%w: failed to load profile: %s", ErrImportFailed, err)
+	}
+
+	// FIXME: check if there are any global-only setting in the import
+
+	// Stop here if we are only validating.
+	if r.ValidateOnly {
+		return result, nil
+	}
+
+	// Import settings into profile.
+	if r.Reset {
+		p.Config = config.Expand(settings)
+	} else {
+		flattenedProfileConfig := config.Flatten(p.Config)
+		for k, v := range settings {
+			flattenedProfileConfig[k] = v
+		}
+		p.Config = config.Expand(flattenedProfileConfig)
+	}
```

Comment:
Same as above, the whole flattening and expanding of config options seems like useless overhead